src/ocean/math/WideUInt.d

```diff
@@ -0,0 +1,591 @@
+            ulong value = this.payload[1];
+            value <<= 32;
+            value |= this.payload[0];
+            return cast(double) value;
+        }
+
+        // else calculate floating point value from 3 most significant words
+        double MAXWORD = pow(2.0, 32.0);
+        return ((this.payload[idx] * MAXWORD + this.payload[idx-1])
+                * MAXWORD + this.payload[idx-2])
+            * ldexp(1.0, (idx-2)*32);
+    }
+
+    unittest
+    {
+        WideUInt num = 123456789;
+        test!("is")(num.toDouble(), 123_456_789.0);
+
+        for (int i = 0; i < 3; ++i)
+            num.multiplyBy(1_000_000_000);
+
+        test(feqrel(num.toDouble(), 123_456_789e+27) == double.mant_dig);
+    }
+
+    /***************************************************************************
+
+        Helper utility to increment word under index `idx`, calling itself
+        recursively for following words if wraparound happens.
+
+        Params:
+            idx = index of word in the internal payload to increment by 1
+
+        Returns:
+            'false' if whole WideUInt has overflown (== error)
+
+    ***************************************************************************/
+
+    private bool checkAndInc ( size_t idx )
+    {
+        if (idx >= N)
+            return false;
+
+        uint after = ++this.payload[idx];
+        if (after == 0)
+            return this.checkAndInc(idx+1);
+
+        return true;
+    }
+
+    unittest
+    {
+        WideUInt num = uint.max;
+        test(num.checkAndInc(0));
+        test(num.payload[0] == 0);
+        test(num.payload[1] == 1);
+
+        num.payload[$-1] = uint.max;
+        test(!num.checkAndInc(num.payload.length - 1));
+    }
+
+    /***************************************************************************
+
+        Trivial helper to initialize to least significant words of the number
+        from ulong.
+
+    ***************************************************************************/
+
+    private void assign ( ulong value )
+    {
+        this.payload[] = 0;
+        this.payload[0] = cast(uint) value;
+        this.payload[1] = cast(uint) (value >> 32);
+    }
+
+    /***************************************************************************
+
+        Helper utility that adds two uint integers checking for wraparound
+
+        Params:
+            a = reference to integer to increment
+            b = number to add
+
+        Returns:
+            'true' if wraparound/overflow has happenned, 'false' otherwise
+
+    ***************************************************************************/
+
+    static private bool add ( ref uint a, uint b )
+    {
+        bool flag;
+
+        version (D_InlineAsm_X86_64)
+        {
+            // When adding support for other compilers, optimized machine
+            // code output should be checked to see if assembly block can be
+            // reliably replaced with plain D arithmetics
+
+            asm
```

Comment:
I would guess this would be pretty close:

```
static uint add ( uint a, uint b, out bool overflow )
{
    ulong res = cast(ulong)(a) + cast(ulong)(b);
    overflow = res >> 0x20;
    return cast(uint)res;
}
```